cbrkit/sim/collections.py

```diff
@@ -26,3 +26,51 @@ def wrapped_func(x: Collection[Any], y: Collection[Any]) -> float:
         return dist2sim(jaccard_distance(x, y))
 
     return wrapped_func
+
+def smith_waterman(match_score: int = 2, mismatch_penalty: int = -1, gap_penalty: int = -1) -> SimPairFunc:
+    """
+    Performs the Smith-Waterman alignment with configurable scoring parameters. If no element matches it returns 0.0.
+
+    Args:
+        match_score (int, optional): Score for matching characters. Defaults to 2.
+        mismatch_penalty (int, optional): Penalty for mismatching characters. Defaults to -1.
+        gap_penalty (int, optional): Penalty for gaps. Defaults to -1.
+
+    Returns:
+        float: Alignment score for the two sequences.
+
+    Example:
+        >>> sim = smith_waterman()
+        >>> sim("abcde", "fghe")
+        2.0
+    """
+    from minineedle import smith, core
+
+    def wrapped_func(x: str, y: str) -> float:
+        try:
+            alignment = smith.SmithWaterman(x, y)
+            alignment.change_matrix(core.ScoreMatrix(match=match_score, miss=mismatch_penalty, gap=gap_penalty))
+            alignment.align()
+            return alignment.get_score()
+        except ZeroDivisionError:
+            return 0.0
+
+    return wrapped_func
+def dtw_similarity() -> SimPairFunc:
+    """Dynamic Time Warping similarity function.
+
+    Examples:
+        >>> sim = dtw_similarity()
+        >>> sim([1, 2, 3], [1, 2, 3, 4])
+        # Wert hängt von der konkreten Implementierung der Umrechnung von Distanz zu Ähnlichkeit ab
```

Comment:
Does this mean the similarity computations is not deterministic?

Reply:
There are different ways to go from a distance measure to a similarity measure (eg gaussian), so the similarity here depends on how this conversion is done.

Reply:
Ok, makes sense. But where would this be configured? It should (based on my understanding) not cause issues in this example. Or did I not get something important here?

Reply:
Now that it uses the helper function such a distance to similarity conversion would be the same over all functions that use it, so that gives consistency and therefore no problems for this function. It could cause an issue if someone expects a different conversion... maybe it would help to add an optional boolean to the parameters of the function to either give the similarity with the given conversion (default) or it gives the distance and a user could write a custom conversion?